### PR TITLE
Adding upload_file adapter

### LINKED ISSUE #102: Add macro/run-operation for uploading files from local filesystem to BigQuery.
### Describe the feature
Prompted by https://github.com/brooklyn-data/dbt_artifacts/issues/6.

Access to a macro such as:

{% do adapter.upload_file(file_path, destination) %}
would enable dbt_artifacts to become compatible with BigQuery. It looks as though the function to wrap is [load_table_from_file](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.load_table_from_file).

### Additional context
Please include any other relevant context here.

### Who will this benefit?
What kind of use case will this feature be useful for? Please be specific and provide examples, this will help us prioritize properly.

### Are you interested in contributing this feature?
Let us know if you want to write some code, and how we can help.

### PR DESCRIPTION
**Pull request by pgoslatara**
resolves #102

### Description

Adding an `upload_file` adapter that uses the [load_table_from_file](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.load_table_from_file) method. This adapter takes multiple arguments allowing maximum customisation of the [LoadJobConfig](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.LoadJobConfig.html#google.cloud.bigquery.job.LoadJobConfig) class.

Example files:
- ./samples/three_columns.csv:
```
id,name,created_at
1,john,2022-02-06 19:24:23.614 UTC
```
- ./samples/id_only.csv:
```
id
1
```
- ./samples/users.ndjson:
```
{"id":1,"name":"Alice"}
{"id":2,"name":"Bob"}
{"id":3,"name":"Carol"}
```

How to use:
- Via command line:
```
dbt -d run-operation bigquery__upload_file --args "{local_file_path: './samples/three_columns.csv', database: '<GCP_PROJECT>', table_schema: 'dbt', table_name: 'three_columns', skip_leading_rows: 1, autodetect: True, write_disposition: 'WRITE_TRUNCATE'}"

dbt -d run-operation bigquery__upload_file --args "{local_file_path: './samples/id_only.csv', database: '<GCP_PROJECT>', table_schema: 'dbt', table_name: 'upload_id', skip_leading_rows: 1, write_disposition: 'WRITE_TRUNCATE', schema: '[{\"name\": \"employee_id\", \"type\": \"INTEGER\"}]'}"

dbt -d run-operation bigquery__upload_file --args "{local_file_path: './samples/users.ndjson', database: '<GCP_PROJECT>', table_schema: 'dbt', table_name: 'users', autodetect: True, write_disposition: 'WRITE_TRUNCATE', source_format: 'NEWLINE_DELIMITED_JSON'}"
```
- Via a pre_hook:
```
{{
    config(
        pre_hook=[
            "{{ bigquery__upload_file(
                './samples/three_columns.csv',
                model['database'],
                model['schema'],
                'three_columns',
                autodetect=True,
                skip_leading_rows=1,
                write_disposition='WRITE_TRUNCATE',
                source_format='CSV'
            ) }}",
            "{{ bigquery__upload_file(
                './samples/id_only.csv',
                model['database'],
                model['schema'],
                'id_only',
                skip_leading_rows=1,
                write_disposition='WRITE_TRUNCATE',
                source_format='CSV',
                schema='[{\"name\": \"employee_id\", \"type\": \"INTEGER\"}]'
            ) }}",
            "{{ bigquery__upload_file(
                './samples/users.ndjson',
                model['database'],
                model['schema'],
                'users',
                autodetect=True,
                write_disposition='WRITE_TRUNCATE',
                source_format='NEWLINE_DELIMITED_JSON'
            ) }}"
        ]
    )
}}
```

Open questions from my side:
- Currently the adapter only logs the kwargs, do we want to log something else?
- The most similar adapter is `load_dataframe` which does not have tests. Should an issue be opened to add tests for both `load_dataframe` and `upload_file`?

Comment:
Neither `manifest.json` or `run_results.json` can be uploaded using this adapter as they do not conform to ndjson specifications. If this PR is merged an issue can be opened to address this (possibly alter these files before calling the `upload_file` adapter).

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-bigquery next" section.

### CODE REVIEW

**Comment by McKnight-42** [comment]
Hi @pgoslatara taking a look over today and will pass along some questions sometime tomorrow, sorry for the delay.

File: dbt/adapters/bigquery/impl.py [hunk 1]
```diff
@@ -686,6 +686,29 @@ def load_dataframe(self, database, schema, table_name, agate_table,
...
+            if k == "schema":
+                setattr(load_config, k, json.loads(v))
+            else:
+                setattr(load_config, k, v)
+
+        with open(local_file_path, "rb") as f:
+            job = client.load_table_from_file(f, table_ref, rewind=True,
+                                              job_config=load_config)
+
+        timeout = self.connections.get_timeout(conn)
```
Comment (McKnight-42) [inline dbt/adapters/bigquery/impl.py:708]: @pgoslatara this `get_timeout` method was throwing an error for me when testing ex:`'BigQueryConnectionManager' object has no attribute 'get_timeout'` and was curious if you had a package installed that gave you access to it? I was able to replace it with the `get_job_execution_timeout_seconds` method as seen in `load_dataframe`

**Comment by McKnight-42** [comment]
> Currently the adapter only logs the kwargs, do we want to log something else?
@jtcohen6 I'm curious if you have any thoughts on this one way or the other?

**Comment by McKnight-42** [comment]
> Should an issue be opened to add tests for both load_dataframe and upload_file?

@pgoslatara good catch on the missing `load_dataframe` test. I will open an issue to create a test for that.

 For the `upload_file` test I think we should try to if able write up a test along with this PR.

File: dbt/adapters/bigquery/impl.py (hunk 1 above)
Comment (pgoslatara) [reply dbt/adapters/bigquery/impl.py]: Looks like my branch was outdated, it was missing this [commit](https://github.com/dbt-labs/dbt-bigquery/commit/141b86749df813cf3a3a90a90e7a7dfc401ba9b0). I've updated this method as you suggested, [commit](https://github.com/dbt-labs/dbt-bigquery/pull/121/commits/7b23a10890567cbad2cc4c4367c54f122895c24b).

**Comment by pgoslatara** [comment]
> For the `upload_file` test I think we should try to if able write up a test along with this PR.

@McKnight-42 I've added a basic test for uploading three different types of files (CSV, NDJSON and parquet), bfe11ba1860015af73ac09379acd913d65ef3ce1. I've never previously worked with tests so this is all new to me. Can you take a look and see if these make sense?

File: tests/integration/upload_file_tests/test_upload_file.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+from tests.integration.base import DBTIntegrationTest, use_profile
+import yaml
+
+
+class TestUploadFileCSV(DBTIntegrationTest):
+    @property
+    def schema(self):
+        return "adapter_methods_caching"
```
Comment (McKnight-42) [inline tests/integration/upload_file_tests/test_upload_file.py:8]: I might suggest a name change on here and lines 32 and 56 to something like 'upload_file'.

**Comment by McKnight-42** [comment]
@pgoslatara Great start on the tests though I feel like we need to go a little further, possibly adding calls to the newly created tables and to do some simple check that they actually have some information just incase a error happens and database only creates a empty table.

File: tests/integration/upload_file_tests/test_upload_file.py (hunk 2 above)
Comment (pgoslatara) [reply tests/integration/upload_file_tests/test_upload_file.py]: Good spot! Done, 1f03a041047e6820c6d717807dbf5b9111a6104b.

**Comment by pgoslatara** [comment]
This is a great suggestion.

I've added checks for the number of rows, distinct `id` values, maximum `updated_at` value and that the data types are as expected, 1ed8a9d2e085e74aebbc04bc0dee9a00eae4f965. I've done a bit of refactoring here so as to avoid having to define the checks for each uploaded table, let me know if this doesn't make sense or has some downsides. Also had to re-create the parquet file so `updated_at` is a timestamp rather than a string.

**Comment by pgoslatara** [comment]
@McKnight-42 Done!

Thanks for your input on this one, I learned a lot about how to think about tests.

**Review by McKnight-42 — APPROVED** [review]
Tests both success of load into database and pull down of table information to make sure its all there, covers many forms of upload, well done!

**Pull request merged**